File: 2021/csharp/Aoc2021/Day04.cs
namespace Aoc2021;

public class Day04 : IDay
{
    public override string Day { get; } = nameof(Day04)[3..];

    public override string Result1()
    {
        return Game
            .FromLines(InputLines)
            .PlayWinningBoard()
            .Score
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return Game
            .FromLines(InputLines)
            .PlayLoosingBoard()
            .Score
            .ToString(CultureInfo.InvariantCulture);
    }

    public record Game(List<int> Numbers, List<Board> Boards)
    {
        public Board PlayWinningBoard()
        {
            foreach (int number in Numbers)
            {
                foreach (Board board in Boards)
                {
                    if (board.Play(number))
                    {
                        return board;
                    }
                }
            }

            throw new ApplicationException("No Winner!");
        }

        public Board PlayLoosingBoard()
        {
            Board? lastWinningBoard = null;

            foreach (int number in Numbers)
            {
                List<int> winners = new();
                for (int i = 0; i < Boards.Count; i++)
                {
                    Board board = Boards[i];
                    if (board.Play(number))
                    {
                        winners.Add(i);
                        lastWinningBoard = board;
                    }
                }

                winners.Reverse();
                foreach (int i in winners)
                {
                    Boards.RemoveAt(i);
                }
            }

            return lastWinningBoard ?? throw new ApplicationException("No Winner!");
        }

        public static Game FromLines(IEnumerable<string> lines)
        {
            List<int> numbers = lines.First().Split(',').Select(ParseInt).ToList();
            List<Board> boards =
                lines.Skip(1)
                .Chunk(6)
                .Select(chunk => Board.FromLines(chunk.Skip(1)))
                .ToList();
            return new(numbers, boards);
        }
    }

    public record Board(List<Cell> Cells)
    {
        public int[] RowCounts { get; } = { 0, 0, 0, 0, 0 };
        public int[] ColumnCounts { get; } = { 0, 0, 0, 0, 0 };
        public int Score { get; private set; }

        public bool Play(int number)
        {
            int index = Cells.FindIndex(cell => cell.Value == number);
            if (index < 0)
            {
                return false;
            }

            Cells[index] = Cells[index] with { Marked = true };

            int row = index / 5;
            RowCounts[row] += 1;
            int column = index % 5;
            ColumnCounts[column] += 1;
            bool bingo = RowCounts[row] == 5 || ColumnCounts[column] == 5;
            if (bingo)
            {
                int sum = Cells.Sum(cell => cell.Marked ? 0 : cell.Value);
                Score = number * sum;
            }

            return bingo;
        }

        public static Board FromLines(IEnumerable<string> lines)
        {
            List<Cell> cells =
                lines
                .SelectMany(line => line
                    .Split(' ', StringSplitOptions.RemoveEmptyEntries)
                    .Select(value => new Cell(ParseInt(value))))
                .ToList();
            return new(cells);
        }
    }

    public record Cell(int Value, bool Marked = false);
}

File: 2021/fsharp/Aoc2021/Day04.fs
module Day04

open Utils

type Cell = { Number: int; Marked: bool }

type Board = { Cells: Cell list; Score: int option }

module Board =
    let ofLines lines =
        let cells =
            lines
            |> List.map (String.parseInts ' ')
            |> List.concat
            |> List.map (fun number -> { Number = number; Marked = false })

        { Cells = cells; Score = None }

    let score board =
        match board.Score with
        | Some score -> score
        | None -> 0

    let play n board =
        let cells =
            board.Cells
            |> List.map (fun cell ->
                match cell with
                | { Number = x } when x = n -> { cell with Marked = true }
                | _ -> cell)

        let rows = cells |> List.chunkBySize 5
        let colums = rows |> List.transpose

        let isAllMarked cells =
            cells |> List.forall (fun cell -> cell.Marked)

        let bingo =
            rows |> List.exists isAllMarked || colums |> List.exists isAllMarked

        let score =
            match bingo with
            | false -> None
            | true ->
                let sumOfUnmarked =
                    cells
                    |> List.sumBy (fun cell ->
                        if cell.Marked then 0 else cell.Number)

                n * sumOfUnmarked |> Some

        { Cells = cells; Score = score }

type Game =
    { Numbers: int list; Boards: Board seq }

module Game =
    let ofLines lines =
        let numbers =
            lines
            |> List.head
            |> String.split ','
            |> Array.toList
            |> List.map int

        let boards =
            lines
            |> List.tail
            |> List.chunkBySize 6
            |> List.map (fun chunk -> chunk |> List.tail |> Board.ofLines)

        { Numbers = numbers; Boards = boards }

    let rec play game =
        seq {
            let n = game.Numbers |> List.head
            let boards = game.Boards |> Seq.map (Board.play n)

            let winners =
                boards |> Seq.filter (fun board -> board.Score |> Option.isSome)

            yield! winners

            let loosers =
                boards |> Seq.filter (fun board -> board.Score |> Option.isNone)

            let numbers = game.Numbers |> List.tail

            if Seq.isEmpty loosers || List.isEmpty numbers then
                yield! Seq.empty
            else
                yield! play { Numbers = numbers; Boards = loosers }
        }

let input = "04" |> readInputLines |> Seq.toList

let job1 () =
    input |> Game.ofLines |> Game.play |> Seq.head |> Board.score |> string

let job2 () =
    input |> Game.ofLines |> Game.play |> Seq.last |> Board.score |> string
